### ISSUE TITLE
Crashes with errors when trying to run huey_consumer with "process" worker type on Microsoft Windows 7

### ISSUE THREAD

**Issue by ctsrc** [coleifer/huey#306]
Hi, I am using huey and it's been very good so far.

I now need to execute multiple CPU-heavy tasks in parallel.

http://huey.readthedocs.io/en/latest/consumer.html says:

> `-k`, `--worker-type`
> Choose the worker type, `thread`, `process` or `greenlet`. The default is `thread`.
> 
> Depending on your workload, one worker type may perform better than the others:
> 
> * CPU heavy loads: use "process". Python's global interpreter lock prevents multiple threads from running simultaneously, so to leverage multiple CPU cores (and reduce thread contention) run each worker as a separate process.
> * IO heavy loads: use "greenlet". For example, tasks that crawl websites or which spend a lot of time waiting to read/write to a socket, will get a huge boost from using the greenlet worker model. Because greenlets are so cheap in terms of memory, you can easily run tens or hundreds of them.
> * Anything else: use "thread". You get the benefits of pre-emptive multi-tasking without the overhead of multiple processes. A safe choice and the default.

However when I tried to run two worker processes for my application with `.\huey_consumer.exe -k process -w 2 cne_tool.webui.tasks.huey`, it crashes with the following output:

    [2018-03-31 21:48:33,815] INFO:huey.consumer:8196:Huey consumer started with 2 process, PID 8196
    [2018-03-31 21:48:33,816] INFO:huey.consumer:8196:Scheduler runs every 1 second(s).
    [2018-03-31 21:48:33,817] INFO:huey.consumer:8196:Periodic tasks are enabled.
    [2018-03-31 21:48:33,817] INFO:huey.consumer:8196:UTC is enabled.
    [2018-03-31 21:48:33,818] INFO:huey.consumer:8196:The following commands are available:
    + queue_task_import_rawdata
    + queue_task_make_xml_allsteps
    Traceback (most recent call last):
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\runpy.py", line 193, in _run_module_as_main
        "__main__", mod_spec)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\runpy.py", line 85, in _run_code
        exec(code, run_globals)
      File "C:\Users\erikn\AppData\Local\Statnett CNE Tool\Scripts\huey_consumer.exe\__main__.py", line 9, in <module>
      File "c:\users\erikn\appdata\local\statnett cne tool\lib\site-packages\huey\bin\huey_consumer.py", line 47, in consumer_main
        consumer.run()
      File "c:\users\erikn\appdata\local\statnett cne tool\lib\site-packages\huey\consumer.py", line 535, in run
        self.start()
      File "c:\users\erikn\appdata\local\statnett cne tool\lib\site-packages\huey\consumer.py", line 512, in start
        self.scheduler.start()
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\process.py", line 105, in start
        self._popen = self._Popen(self)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\context.py", line 223, in _Popen
        return _default_context.get_context().Process._Popen(process_obj)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\context.py", line 322, in _Popen
        return Popen(process_obj)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\popen_spawn_win32.py", line 65, in __init__
        reduction.dump(process_obj, to_child)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\reduction.py", line 60, in dump
        ForkingPickler(file, protocol).dump(obj)
    AttributeError: Can't pickle local object 'Consumer._create_process.<locals>._run'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\spawn.py", line 99, in spawn_main
        new_handle = reduction.steal_handle(parent_pid, pipe_handle)
      File "C:\Users\erikn\AppData\Local\Python-3.6.4\lib\multiprocessing\reduction.py", line 87, in steal_handle
        _winapi.DUPLICATE_SAME_ACCESS | _winapi.DUPLICATE_CLOSE_SOURCE)
    PermissionError: [WinError 5] Ingen tilgang

**Comment by coleifer** [comment]
I guess you won't be able to run multiprocess mode on windows. Try threads? Wontfix.

**Comment by ctsrc** [comment]
I was using threads first but because the task is long running and CPU heavy it blocks all other tasks from running until it has completed.

While you don't want to fix this issue yourself would you accept pull requests if someone else were to fix it? Don't have time to do so myself atm but might in the future if I know that PR would be accepted.

**Comment by coleifer** [comment]
The interpreter will preempt threads, so even though you're effectively running single-threaded, they should switch every x instructions.

**Comment by ctsrc** [comment]
This task takes several minutes to complete and only a very little amount of it is I/O and it does all the reading up front and all the writing at the end so once it's running it runs till completion which is why only multi-process parallelism will help me.

**Comment by ctsrc** [comment]
What would happen if I were to run multiple instances of huey_consumer side by side? Would that be safe and give me parallel execution, or could race-conditions occur such that multiple consumer instances start the same task, or would it just lead to straight up corruption right off the bat?

**Comment by ctsrc** [comment]
It turned out to be not quite like I said -- it does switch a little between tasks when threaded. Still it is CPU heavy and it ends up taking a lot of time. Even more so when I have multiple of that kind of task running of course and it's a shame to not be able to utilize the multiple CPU threads which could have reduced the time taken by a lot. The last question I asked still stands though wrt what would happen if I were to run multiple instances of huey_consumer side by side.

**Comment by coleifer** [comment]
You should be fine to run multiple huey instances side-by-side. It strikes me as odd that you are running huey on Windows for your data-processing.

**Comment by ctsrc** [comment]
I am making this software for someone else. Personally I don’t run Windows on any of my computers. The choice for the server that this will run on to be running Windows Server was not a choice of my own.

Personally I think Windows is a horrible experience in every way possible.

Anyway I landed on Huey because it seemed nice to use and sufficiently documented. The other alternatives I looked at either stated that they didn’t support Windows (for example Celery said this) or they seemed too convoluted or too lacking in documentation.

Some parts of Huey were a bit difficult to understand and I struggled a bit with it but all over I am glad that Huey exists and that I chose to use it.

Glad to hear that running multiple Huey instances side-by-side should be fine. Thank you.
